src/theory/quantifiers/expr_miner_manager.h

```diff
@@ -0,0 +1,103 @@
+
+#ifndef __CVC4__THEORY__QUANTIFIERS__EXPR_MINER_MANAGER_H
+#define __CVC4__THEORY__QUANTIFIERS__EXPR_MINER_MANAGER_H
+
+#include "theory/quantifiers/candidate_rewrite_database.h"
+#include "theory/quantifiers/extended_rewrite.h"
+#include "theory/quantifiers/sygus_sampler.h"
+#include "theory/quantifiers_engine.h"
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+/** ExpressionMinerManager
+ *
+ * This class manages a set of expression miners. It provides a common place
+ * to register expression so that multiple mining algorithms can be run in
+ * coordination, possibly sharing information and utilities like a common
+ * sampling object.
+ */
+class ExpressionMinerManager
+{
+ public:
+  ExpressionMinerManager();
+  ~ExpressionMinerManager() {}
+  /**  Initialize this class
+   *
+   * Initializes this class, informing it that the free variables of terms
+   * added to this class via addTerm will have free variables that are a subset
+   * of vars, and have type tn. The arguments nsamples and unique_type_ids are
+   * used for initializing the sampler class of this manager
+   * (see SygusSampler::initialize for details).
+   */
+  void initialize(const std::vector<Node>& vars,
+                  TypeNode tn,
+                  unsigned nsamples,
+                  bool unique_type_ids = false);
+  /** Initialize this class, sygus version
+   *
+   * Initializes this class, informing it that the terms added to this class
+   * via calls to addTerm will be generated by the grammar of f. The method
+   * takes a pointer to the quantifiers engine qe. If the argument useSygusType
+   * is true, the terms added to this class are the sygus datatype terms.
+   * If useSygusType is false, the terms are the builtin equivalent of these
+   * terms. The argument nsamples is used to initialize the sampler.
+   */
+  void initializeSygus(QuantifiersEngine* qe,
+                       Node f,
+                       unsigned nsamples,
+                       bool useSygusType);
+  /** enable rewrite rule synthesis (--sygus-rr-synth) */
+  void enableRewriteRuleSynth();
+  /** add term
+   *
+   * Expression miners may print information on the output stream out, for
+   * instance, candidate-rewrites.
+   */
+  bool addTerm(Node sol, std::ostream& out);
+  /**
+   * Same as above, but the argument rew_print is set to true if a rewrite rule
+   * was printed on the output stream out.
+   */
+  bool addTerm(Node sol, std::ostream& out, bool& rew_print);
+
+ private:
+  /** whether we are doing rewrite synthesis */
+  bool d_do_rew_synth;
+  /** the sygus function passed to initializeSygus, if any */
+  Node d_sygus_fun;
+  /** whether we are using sygus types */
+  bool d_use_sygus_type;
+  /** pointer to the quantifiers engine, used if d_use_sygus is true */
+  QuantifiersEngine* d_qe;
+  /** the sygus term database of d_qe */
+  TermDbSygus* d_tds;
+  /** candidate rewrite database */
+  CandidateRewriteDatabase d_crd;
```

Comment:
Shouldn't this be an `ExprMiner*` or a list of `ExprMiner`s?

Reply:
At the moment, no. The idea is that this class is going to manage the interaction between expression miners, which is unique to the expression miners in question. For example, if query generation is enabled, then rewrite rule generation is used to filter query generation. This interaction is something that the individual miners should not have to worry about (so it can't be virtual-ized).

Reply:
Hmm, the reason why I am asking is that we are currently not really using the fact that `CandidateRewriteDatabase` is an `ExprMiner`. If we are planning to treat all the `ExprMiner`s as specialized objects, then maybe we should consider not even having an `ExprMiner` super class and just move `initializeChecker` et al. to some kind of utils class. Maybe we can reconsider after we have a couple of `ExprMiner`s.

Reply:
Well, there are still common data members: the variables and sampler.

Reply:
An alternative design would be to package the variables and the sampler into a class that is then a member variable of the different `ExprMiner`s. What's better is rather subjective.